cumulusci/oauth/client.py

```diff
@@ -0,0 +1,293 @@
+            "redirect_uri": self.client_info.callback_url,
+            "code": auth_code,
+        }
+        return requests.post(
+            self.client_info.token_uri, headers=HTTP_HEADERS, data=data
+        )
+
+    def refresh_token(self, refresh_token):
+        data = {
+            "client_id": self.client_info.client_id,
+            "client_secret": self.client_info.client_secret,
+            "grant_type": "refresh_token",
+            "refresh_token": refresh_token,
+        }
+        response = requests.post(
+            self.client_info.token_uri, headers=HTTP_HEADERS, data=data
+        )
+        self.validate_response(response)
+        return safe_json_from_response(response)
+
+    def revoke_token(self, access_token):
+        data = {"token": quote(access_token)}
+        revoke_uri = self.client_info.revoke_uri
+        response = requests.post(revoke_uri, headers=HTTP_HEADERS, data=data)
+        response.raise_for_status()
+        return response
+
+    def validate_response(self, response):
+        """Subclasses can implement custom response validation"""
+        if not response:
+            raise CumulusCIUsageError("Authentication timed out or otherwise failed.")
+        elif response.status_code == http.client.OK:
+            return
+        raise OAuthError(
+            f"OAuth failed\nstatus_code: {response.status_code}\ncontent: {response.content}"
+        )
+
+
+class HTTPDTimeout(threading.Thread):
+    """Establishes a timeout for a SimpleHTTPServer"""
+
+    # allow the process to quit even if the
+    # timeout thread is still alive
+    daemon = True
+
+    def __init__(self, httpd, timeout):
+        self.httpd = httpd
+        self.timeout = timeout
+        super().__init__()
+
+    def run(self):
+        """Check every second for HTTPD or quit after timeout"""
+        target_time = time.time() + self.timeout
+        while time.time() < target_time:
+            time.sleep(1)
+            if not self.httpd:
+                break
+
+        # extremely minor race condition
+        if self.httpd:  # pragma: no cover
+            self.httpd.shutdown()
+
+    def quit(self):
+        """Quit before timeout"""
+        self.httpd = None
+
+
+class OAuthCallbackHandler(BaseHTTPRequestHandler):
+    parent = None
+
+    def do_GET(self):
+        args = parse_qs(urlparse(self.path).query, keep_blank_values=True)
+
+        if "error" in args:
+            http_status = http.client.BAD_REQUEST
+            http_body = f"error: {args['error'][0]}\nerror description: {args['error_description'][0]}"
+        else:
+            http_status = http.client.OK
+            emoji = random.choice(["🎉", "👍", "👍🏿", "🥳", "🎈"])
+            http_body = f"""<html>
+            <h1 style="font-size: large">{emoji}</h1>
+            <p>Congratulations! Your authentication succeeded.</p>"""
+            auth_code = args["code"]
+            self.parent.response = self.parent.get_access_token(auth_code)
```

Comment:
Here is where we need a special case to handle the tssd from Marketing Cloud. If the MC installed package is a partner package, we'll get tssd in `args` here, and need to add it to the response stored on self.parent so that it gets stored in the service config.